lightwallet-client-lib/src/main/java/co/electriccoin/lightwallet/client/internal/BlockingLightWalletClientImpl.kt

```diff
@@ -0,0 +1,191 @@
+ *
+ * @property channel the channel to use for communicating with the lightwalletd server.
+ * @property singleRequestTimeout the timeout to use for non-streaming requests. When a new stub
+ * is created, it will use a deadline that is after the given duration from now.
+ * @property streamingRequestTimeout the timeout to use for streaming requests. When a new stub
+ * is created for streaming requests, it will use a deadline that is after the given duration from
+ * now.
+ */
+internal class BlockingLightWalletClientImpl private constructor(
+    private val channelFactory: ChannelFactory,
+    private val lightWalletEndpoint: LightWalletEndpoint,
+    private val singleRequestTimeout: Duration = 10.seconds,
+    private val streamingRequestTimeout: Duration = 90.seconds
+) : BlockingLightWalletClient {
+
+    private var channel = channelFactory.newChannel(lightWalletEndpoint)
+
+    override fun getBlockRange(heightRange: ClosedRange<BlockHeightUnsafe>): Sequence<CompactFormats.CompactBlock> {
+        require(!heightRange.isEmpty()) {
+            "${Constants.ILLEGAL_ARGUMENT_EXCEPTION_MESSAGE} range: $heightRange." // NON-NLS
+        }
+
+        return requireChannel().createStub(streamingRequestTimeout)
+            .getBlockRange(heightRange.toBlockRange()).iterator().asSequence()
+    }
+
+    override fun getLatestBlockHeight(): Response<BlockHeightUnsafe> {
+        return try {
+            if (BenchmarkingExt.isBenchmarking()) {
+                // We inject a benchmark test blocks range at this point to process only a restricted range of blocks
+                // for a more reliable benchmark results.
+                Response.Success(BlockHeightUnsafe(BlockRangeFixture.new().endInclusive))
+            } else {
+                val response = requireChannel().createStub(singleRequestTimeout)
+                    .getLatestBlock(Service.ChainSpec.newBuilder().build())
+
+                val blockHeight = BlockHeightUnsafe(response.height)
+
+                Response.Success(blockHeight)
+            }
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    @Suppress("SwallowedException")
+    override fun getServerInfo(): Response<LightWalletEndpointInfoUnsafe> {
+        return try {
+            val lightdInfo = requireChannel().createStub(singleRequestTimeout)
+                .getLightdInfo(Service.Empty.newBuilder().build())
+
+            val lightwalletEndpointInfo = LightWalletEndpointInfoUnsafe.new(lightdInfo)
+
+            Response.Success(lightwalletEndpointInfo)
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    override fun submitTransaction(spendTransaction: ByteArray): Response<SendResponseUnsafe> {
+        if (spendTransaction.isEmpty()) {
+            return Response.Failure.Client.SubmitEmptyTransaction()
```

Comment:
Long-term, is this and `fetchTransaction` below instances of where we should be throwing an Exception?

Reply:
Yep, I just processed to right away. The exception throwing approach is better for us in this state of refactoring.